File: schemas/vyuh-sanity-schema-auth/src/content/email-password-form.ts
import { defineField, defineType, SchemaTypeDefinition } from 'sanity';
import { SiAuth0 as Icon } from 'react-icons/si';
import {
  ContentDescriptor,
  ContentSchemaBuilder,
} from '@vyuh/sanity-schema-core';
import { showLoginErrorField } from './util';

export class EmailPasswordFormDescriptor extends ContentDescriptor {
  static schemaType = 'auth.emailPasswordForm';

  constructor(props: Partial<EmailPasswordFormDescriptor>) {
    super(EmailPasswordFormDescriptor.schemaType, props);
  }
}

export class EmailPasswordFormSchemaBuilder extends ContentSchemaBuilder {
  private schema = defineType({
    name: EmailPasswordFormDescriptor.schemaType,
    title: 'Email Password Form',
    type: 'object',
    icon: Icon,
    fields: [
      defineField({
        name: 'actionType',
        title: 'Auth Action Type',
        description: 'The type of authentication action to perform',
        type: 'string',
        initialValue: 'signIn',
        options: {
          list: [
            { title: 'Sign In', value: 'signIn' },
            { title: 'Sign Up', value: 'signUp' },
          ],
        },
      }),
      defineField({
        name: 'showPasswordVisibilityToggle',
        title: 'Show Password Visibility Toggle',
        type: 'boolean',
        initialValue: true,
      }),
      showLoginErrorField(),
      defineField({
        name: 'action',
        title: 'Action',
        description:
          'Action to invoke after signing in or after registering a new user',
        type: 'vyuh.action',
      }),
      defineField({
        name: 'forgotPasswordAction',
        title: 'Forgot Password Action',
        description:
          'Action to invoke when user requests to reset their password',
        type: 'vyuh.action',
        hidden: (context) => context.parent.actionType !== 'signIn',
      }),
      defineField({
        name: 'signupAction',
        title: 'Sign Up Action',
        description:
          'Action to invoke when user wants to sign up for a new account',
        type: 'vyuh.action',
        hidden: (context) => context.parent.actionType !== 'signIn',
      }),
      defineField({
        name: 'loginAction',
        title: 'Login Action',
        description:
          'Action to invoke when user wants to login to an existing account',
        type: 'vyuh.action',
        hidden: (context) => context.parent.actionType !== 'signUp',
      }),
    ],
    preview: {
      select: {
        showPasswordVisibilityToggle: 'showPasswordVisibilityToggle',
        showLoginError: 'showLoginError',
      },
      prepare: (props) => {
        return {
          title: 'Email Password Form',
          subtitle: `Show Password Visibility Toggle: ${props.showPasswordVisibilityToggle} | Show Login Error: ${props.showLoginError}`,
        };
      },
    },
  });

  constructor() {
    super(EmailPasswordFormDescriptor.schemaType);
  }

  build(descriptors: ContentDescriptor[]): SchemaTypeDefinition {
    return this.schema;
  }
}

File: packages/system/vyuh_feature_auth/lib/content/email_password_form.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:vyuh_core/vyuh_core.dart';
import 'package:vyuh_extension_content/vyuh_extension_content.dart';
import 'package:vyuh_feature_auth/content/email_password_form_layout.dart';
import 'package:vyuh_feature_auth/ui/email_password_view.dart';

part 'email_password_form.g.dart';

@JsonSerializable()
final class EmailPasswordForm extends ContentItem {
  final bool showPasswordVisibilityToggle;
  final AuthActionType actionType;
  final bool showLoginError;
  final Action? action;
  final Action? forgotPasswordAction;
  final Action? signupAction;
  final Action? loginAction;

  static const schemaName = 'auth.emailPasswordForm';
  static final typeDescriptor = TypeDescriptor(
    schemaType: schemaName,
    title: 'Email Password Form',
    fromJson: EmailPasswordForm.fromJson,
  );
  static final contentBuilder = ContentBuilder(
    content: typeDescriptor,
    defaultLayoutDescriptor: DefaultEmailPasswordFormLayout.typeDescriptor,
    defaultLayout: DefaultEmailPasswordFormLayout(),
  );

  EmailPasswordForm({
    this.showPasswordVisibilityToggle = false,
    this.actionType = AuthActionType.signIn,
    this.showLoginError = true,
    this.action,
    this.forgotPasswordAction,
    this.signupAction,
    this.loginAction,
    super.layout,
  }) : super(schemaType: schemaName);

  factory EmailPasswordForm.fromJson(Map<String, dynamic> json) =>
      _$EmailPasswordFormFromJson(json);
}

final class EmailPasswordFormDescriptor extends ContentDescriptor {
  EmailPasswordFormDescriptor({super.layouts})
      : super(
          schemaType: EmailPasswordForm.schemaName,
          title: 'Email Password Form',
        );
}
